services/leader/client.go

```diff
@@ -0,0 +1,220 @@
+	ec, err := election.NewClient(cli, pfx, election.WithSessionOptions(sessionOpts...))
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		ecl:     ec,
+		opts:    opts.ElectionOpts(),
+		resignC: make(chan struct{}),
+	}, nil
+}
+
+func (c *client) campaign(opts services.CampaignOptions) (<-chan campaign.Status, error) {
+	if c.isClosed() {
+		return nil, errClientClosed
+	}
+
+	if !c.startCampaign() {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c.Lock()
+	c.ctxCancel = cancel
+	c.Unlock()
+
+	proposeVal := c.val(opts.LeaderValue())
+
+	// buffer 1 to not block initial follower update
+	sc := make(chan campaign.Status, 1)
+
+	sc <- campaign.NewStatus(campaign.Follower)
+
+	go func() {
+		defer func() {
+			close(sc)
+			cancel()
+			c.resetCampaign()
+		}()
+
+		// Campaign blocks until elected. Once we are elected, we get a channel
+		// that's closed if our session dies.
+		ch, err := c.ecl.Campaign(ctx, proposeVal)
+		if err != nil {
+			sc <- campaign.NewErrCampaignStatus(err)
+			return
+		}
+
+		sc <- campaign.NewStatus(campaign.Leader)
+		select {
+		case <-ch:
+			sc <- campaign.NewErrCampaignStatus(election.ErrSessionExpired)
+		case <-c.resignC:
+			sc <- campaign.NewStatus(campaign.Follower)
+		}
+	}()
+
+	return sc, nil
+}
+
+func (c *client) resign() error {
+	if c.isClosed() {
+		return errClientClosed
+	}
+
+	// if there's an active blocking call to Campaign() stop it
+	c.Lock()
+	if c.ctxCancel != nil {
+		c.ctxCancel()
+		c.ctxCancel = nil
+	}
+	c.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.opts.ResignTimeout())
+	defer cancel()
+	if err := c.ecl.Resign(ctx); err != nil {
+		return err
+	}
+
+	c.resetCampaign()
+
+	// if successfully resigned and there was a campaign in Leader state cancel
+	// it
+	select {
+	case c.resignC <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+func (c *client) leader() (string, error) {
+	if c.isClosed() {
+		return "", errClientClosed
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.opts.LeaderTimeout())
+	ld, err := c.ecl.Leader(ctx)
+	cancel()
+	if err == concurrency.ErrElectionNoLeader {
+		return ld, ErrNoLeader
+	}
+	return ld, err
+}
+
+func (c *client) startCampaign() bool {
+	return atomic.CompareAndSwapUint32(&c.campaigning, 0, 1)
+}
+
+func (c *client) resetCampaign() {
+	atomic.StoreUint32(&c.campaigning, 0)
+}
+
+// Close closes the election service client entirely. No more campaigns can be
+// started and any outstanding campaigns are closed.
+func (c *client) close() error {
+	atomic.StoreUint32(&c.closed, 1)
+	return c.ecl.Close()
+}
+
+func (c *client) isClosed() bool {
+	return atomic.LoadUint32(&c.closed) == 1
+}
+
+// val returns the value the leader should propose based on (1) a potentially
+// empty override value and (2) the hostname of the caller (with a fallback to
+// the DefaultValue option).
+func (c *client) val(override string) string {
```

Comment:
Removing this, it was an artifact from before I put the logic in `CampaignOptions` itself